File: swift/159_LongestSubstringWithAtMostTwoDistinctCharacters.cpp
class Solution {
public:
    unordered_map<int, int> letter_counts;
    int lengthOfLongestSubstringTwoDistinct(string s) {
        int trail_idx = 0, char_count = 0, curr_length = 0, max_length = 0;
        for (char c : vector<char>('a', 'A')) {
            for (int i = 0; i < 26; ++i) {
                letter_counts[c + i] = 0;
            }
        }
        for (int i = 0; i < s.size(); ++i) {
            curr_length++;
            if (++letter_counts[s[i] - 'A'] == 1) char_count++;
            while (char_count > 2) {
                curr_length--;
                if (--letter_counts[s[trail_idx++] - 'A'] == 0) char_count--;
            }
            max_length = max(curr_length, max_length);
        }
        return max_length;
    }
};

File: swift/159_LongestSubstringWithAtMostTwoDistinctCharacters.swift
class Solution {
    func lengthOfLongestSubstringTwoDistinct(_ s: String) -> Int {
        var letterCounts = [Character : Int]()
        var currLength = 0, maxLength = 0, distinctCharCount = 0
        var trailIdx = s.startIndex
        for (idx, c) in s.enumerated() {
            currLength += 1
            letterCounts[c] = (letterCounts[c] ?? 0) + 1
            if letterCounts[c] == 1 {
                distinctCharCount += 1
            }
            while distinctCharCount > 2 {
                currLength -= 1
                letterCounts[s[trailIdx]]! -= 1
                if letterCounts[s[trailIdx]] == 0 {
                    distinctCharCount -= 1
                }
                trailIdx = s.index(after: trailIdx)
            }
            maxLength = max(maxLength, currLength)
        }
        return maxLength
    }
}
